### ISSUE TITLE
gen_images is generating identical images for a given seed even with noise_mode='random'

### ISSUE THREAD

**Issue by MoemaMike** [NVlabs/stylegan3#76]
I have recent custom trained stylegan3 networks and when i invoke gen_images.py to generate images for a seed list and specify the truncation value  of 0, all of the generated images are identical to each other.  Yet for the same seeds with non zero values for trunc the images are very different from each other, as one would expect. 
 
I can reproduce this by just invoking the simplest of gen_images command lines. I have not tested for a seed range larger than 30 but i don't see why the results would be any different. 

!python stylegan3/gen_images.py --seeds=1-30 --trunc=0  \
--outdir='/content/drive/MyDrive/out' \
--network='/content/drive/MyDrive/networks/x.pkl'

EDIT: see below, I was misinterpreting this as a change in truncation implementation but actually it seems there may be a different issue, specifically noise_mode='random' doing nothing?

**Comment by YukiSakuma** [comment]
trunc=0 means it will generate the average image of the training data.
![image](https://i.imgur.com/CxOlZNb.png)

**Comment by MoemaMike** [comment]
that may be , but with the first and only network i have so far trained on stylegan3 the umages with trunc=0 are IDENTICAL for all seeds.  Maybe my network is not sufficiently trained yet, but i do not think this is the expected result. Certainly this was not the case with any of the networks i trained with stylegan2.

**Comment by jannehellsten** [comment]
You can check the code on how the truncation value is used inside the network:

https://github.com/NVlabs/stylegan3/blob/main/training/networks_stylegan3.py

If you give truncation_psi a zero value, it will throw away your latents and just use w_avg for all samples.  IOW, it will always give you the same output image.

**Comment by MoemaMike** [comment]
> You can check the code on how the truncation value is used inside the network:
> 
> https://github.com/NVlabs/stylegan3/blob/main/training/networks_stylegan3.py
> 
> If you give truncation_psi a zero value, it will throw away your latents and just use w_avg for all samples. IOW, it will always give you the same output image.

ok, so this is by design then.  You do agree this is different from stylegan2 though?

**Comment by MoemaMike** [comment]
closing as working as designed

**Comment by jannehellsten** [comment]
It should work the same way in StyleGAN2:

https://github.com/NVlabs/stylegan2/blob/master/run_generator.py

```
    print('Generating W vectors...')
    all_seeds = list(set(row_seeds + col_seeds))
    all_z = np.stack([np.random.RandomState(seed).randn(*Gs.input_shape[1:]) for seed in all_seeds]) # [minibatch, component]
    all_w = Gs.components.mapping.run(all_z, None) # [minibatch, layer, component]
    all_w = w_avg + (all_w - w_avg) * truncation_psi # [minibatch, layer, component]
    w_dict = {seed: w for seed, w in zip(all_seeds, list(all_w))} # [layer, component]
```

You can see that `all_w` will equal `w_avg` if `truncation_psi` is zero.

**Comment by MoemaMike** [comment]
> It should work the same way in StyleGAN2:
> 
> https://github.com/NVlabs/stylegan2/blob/master/run_generator.py
> 
> ```
>     print('Generating W vectors...')
>     all_seeds = list(set(row_seeds + col_seeds))
>     all_z = np.stack([np.random.RandomState(seed).randn(*Gs.input_shape[1:]) for seed in all_seeds]) # [minibatch, component]
>     all_w = Gs.components.mapping.run(all_z, None) # [minibatch, layer, component]
>     all_w = w_avg + (all_w - w_avg) * truncation_psi # [minibatch, layer, component]
>     w_dict = {seed: w for seed, w in zip(all_seeds, list(all_w))} # [layer, component]
> ```
> 
> You can see that `all_w` will equal `w_avg` if `truncation_psi` is zero.

hmm, yes.  when i generated the stylegan2 images i had --noise-mode='random'  and there were variations based on noise i suppose. 

i am now regenerating my stylegan3 with --noise-mode='random' yet not seeing any variation based on noise even for non zero truncation. Will investigate further, maybe i am doing something wrong and am not getting random noise for some reason

**Comment by nurpax** [comment]
StyleGAN3 does not have noise inputs. I think this explains what you’re seeing.

**Comment by MoemaMike** [comment]
> StyleGAN3 does not have noise inputs. I think this explains what you’re seeing.

hmm, this is a surprise. The code shows noise_mode as a param to generate_images and this line uses it
img = G(z, label, truncation_psi=truncation_psi, noise_mode=noise_mode)

i am a bit confused here. Is it intended to work in the future? If not , is there an alternative that does something similar? Ireally liked the ability to generate alternative images for a given seed based on different noise

**Comment by MoemaMike** [comment]
reopening to better understand what is up with noise_mode ...

**Comment by jannehellsten** [comment]
You should check the paper on noise inputs: https://nvlabs-fi-cdn.nvidia.com/stylegan3/stylegan3-paper.pdf

It's discussed in more detail there.

The stylegan3 codebase supports both stylegan2 and stylegan3 models so the parameter was kept around but it doesn't do anything on stylegan3.

You can compare 

- https://github.com/NVlabs/stylegan3/blob/main/training/networks_stylegan2.py
- https://github.com/NVlabs/stylegan3/blob/main/training/networks_stylegan3.py

Look for "noise_mode" and you can see that it's used in stylegan2 but not in stylegan3.

**Comment by MoemaMike** [comment]
makes sense, thank you
